File: binary_tree/cc/binary_tree.cc
#include <iostream>
#include "binary_tree.h"
#include "queue_linked_list.h"

namespace mf
{

template <class T>
BinaryTreeNode<T> *BinaryTreeNode<T>::add_left_child(const T &value)
{
  left_ = new BinaryTreeNode<T>(value, this);
  return left_;
}

template <class T>
BinaryTreeNode<T> *BinaryTreeNode<T>::add_right_child(const T &value)
{
  right_ = new BinaryTreeNode<T>(value, this);
  return right_;
}

template <class T>
void BinaryTree<T>::pre_order(BinaryTreeNode<T> *node)
{
  if (node == nullptr)
    return;

  std::cout << node->value_ << " ";
  pre_order(node->left_);
  pre_order(node->right_);
}

template <class T>
void BinaryTree<T>::post_order(BinaryTreeNode<T> *node)
{
  if (node == nullptr)
    return;

  post_order(node->left_);
  post_order(node->right_);
  std::cout << node->value_ << " ";
}

template <class T>
void BinaryTree<T>::in_order(BinaryTreeNode<T> *node)
{
  if (node == nullptr)
    return;

  in_order(node->left_);
  std::cout << node->value_ << " ";
  in_order(node->right_);
}

template <class T>
void BinaryTree<T>::level_order()
{
  Queue<BinaryTreeNode<T> *> queue;
  BinaryTreeNode<T> *current;

  queue.enqueue(root_);

  while (!queue.is_empty()) {
    current = queue.dequeue();

    if (current->left_ != nullptr)
      queue.enqueue(current->left_);

    if (current->right_ != nullptr)
      queue.enqueue(current->right_);

    std::cout << current->value_ << " ";
  }
}

}  // namespace mf

File: data-structures/binary_tree/python/binary_tree.py
#!/usr/bin/env python2

from collections import deque


class BinaryTreeNode(object):
  def __init__(self, value=None, parent=None):
    self._parent = parent
    self._value = value
    self._left = None
    self._right = None

  def add_left_child(self, value):
    self._left = BinaryTreeNode(value, self)
    return self._left

  def add_right_child(self, value):
    self._right = BinaryTreeNode(value, self)
    return self._right

  @property
  def left(self):
    return self._left

  @property
  def right(self):
    return self._right

  @property
  def value(self):
    return self._value


class BinaryTree(object):
  def __init__(self, value):
    self._root = BinaryTreeNode(value, None)

  @property
  def root(self):
    return self._root

  def _pre_order(self, node):
    """Depth-first search"""
    if node is None:
      return

    print node.value,
    self._pre_order(node.left)
    self._pre_order(node.right)

  def pre_order(self):
    """Depth-first search"""
    self._pre_order(self._root)

  def _post_order(self, node):
    """Depth-first search"""
    if node is None:
      return

    self._post_order(node.left)
    self._post_order(node.right)
    print node.value,

  def post_order(self):
    """Depth-first search"""
    self._post_order(self._root)

  def _in_order(self, node):
    """Depth-first search"""
    if node is None:
      return

    self._in_order(node.left)
    print node.value,
    self._in_order(node.right)

  def in_order(self):
    """Depth-first search"""
    self._in_order(self._root)

  def level_order(self):
    """Breadth-first search"""
    queue = deque()
    queue.append(self._root)

    while len(queue) > 0:
      node = queue.popleft()

      if node.left is not None:
        queue.append(node.left)

      if node.right is not None:
        queue.append(node.right)

      print node.value,
